db/db_impl/db_impl.cc

```diff
@@ -2636,10 +2676,15 @@ bool DBImpl::MultiCFSnapshot(
     }
   }
 
+  if (!s.ok()) {
+    sv_cleanup_func();
+  }
+
```

Comment:
Thank you for catching this! While making the change for this, I think the original handling for `!unref_only` actually is not doing what it's supposed to be doing, so I revised those too.

It is currently doing:
` iter.cfd->GetSuperVersion()->Unref();`
while I think it meant to do:
`iter.super_version->Unref();`

The former access a column family's current SuperVersion without locking the db mutex, which is not thread safe. And it could also be unrefing a different SuperVersion from what was acquired originally, had there being a new SuperVersion installed in between. Let me know if this fix looks alright. I'm gonna run our stress test suite for this diff before landing it too.

Reply:
> Thank you for catching this! While making the change for this, I think the original handling for `!unref_only` actually is not doing what it's supposed to be doing, so I revised those too.

Wow, great catch!!! Yeah, even though the mutex gets locked during the last try, it is unlocked before `Unref` is actually called, which leaves a window during which another `SuperVersion` might get installed.

Reply:
> I can add these two line fixes in a separate PR and include some unit tests that actually exercising a race condition too. Let me know if you think it's worth separating it out too.

Yeah, that sounds like a good idea - this looks like the kind of scenario that can be reliably tested only using unit tests and `SyncPoint`s

Reply:
Cool, thank you for helping check this! I reverted these two lines in this PR. I will add a separate PR with some unit tests.